Optimize comparing various RenderStyle properties

smfr opened the pull request:
#### 21fec260eb85e4c360106111b32ac8c42e18f862
<pre>
Optimize comparing various RenderStyle properties
<a href="https://bugs.webkit.org/show_bug.cgi?id=286868">https://bugs.webkit.org/show_bug.cgi?id=286868</a>
<a href="https://rdar.apple.com/144024685">rdar://144024685</a>

Reviewed by Sam Weinig.

We can optimize comparisons of things like column span, border colors, content data etc
by first comparing the pointers (`m_nonInheritedData` etc) before then comparing
the values. This is done by adding some helper functions on RenderStyle.

Also optimize `RenderStyle::changeAffectsVisualOverflow()` by grouping and reordering
comparisons.

* Source/WebCore/rendering/RenderLayerModelObject.cpp:
(WebCore::RenderLayerModelObject::styleDidChange):
* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::RenderStyle::changeAffectsVisualOverflow const):
(WebCore::RenderStyle::scrollSnapDataEquivalent const):
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
(WebCore::RenderStyle::borderImageWidth const):
(WebCore::RenderStyle::contentData const):
(WebCore::RenderStyle::columnSpanEqual const):
(WebCore::RenderStyle::borderIsEquivalentForPainting const):
(WebCore::RenderStyle::contentDataEquivalent const):
(WebCore::RenderStyle::containerTypeAndNamesEqual const):
(WebCore::RenderStyle::scrollPaddingEqual const):
* Source/WebCore/rendering/style/StyleSurroundData.h:
* Source/WebCore/style/StyleChange.cpp:
(WebCore::Style::determineChange):
(WebCore::Style::operator&lt;&lt;):
* Source/WebCore/style/StyleChange.h:

Canonical link: <a href="https://commits.webkit.org/289700@main">https://commits.webkit.org/289700@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3b02d368d73d4eadb6a44046e8003ca604306f5c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/87720 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/7236 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/42104 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/92585 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/38466 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/89771 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/7617 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/15407 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/67724 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/25472 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/90722 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/5811 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/79370 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/48095 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/5598 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/33772 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/37577 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/75996 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/34655 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/94471 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/14888 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/10925 "Found 1 new test failure: ipc/create-media-source-with-invalid-constraints-crash.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/76577 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/15142 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/75226 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/75812 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18645 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/20175 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/18609 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/7855 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/14904 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/20206 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/14648 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/18092 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/16430 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

weinig commented on Source/WebCore/rendering/style/RenderStyle.h:
```diff
@@ -1073,11 +1076,13 @@ class RenderStyle final : public CanMakeCheckedPtr<RenderStyle> {
     ScrollSnapType scrollSnapType() const;
     const ScrollSnapAlign& scrollSnapAlign() const;
     ScrollSnapStop scrollSnapStop() const;
+    bool scrollSnapDataEquivalent(const RenderStyle&) const;
```
What is the difference between the `Equivalent` and `Equal` suffixed functions?

smfr committed:
Optimize comparing various RenderStyle properties
https://bugs.webkit.org/show_bug.cgi?id=286868
rdar://144024685

Reviewed by Sam Weinig.

We can optimize comparisons of things like column span, border colors, content data etc
by first comparing the pointers (`m_nonInheritedData` etc) before then comparing
the values. This is done by adding some helper functions on RenderStyle.

Also optimize `RenderStyle::changeAffectsVisualOverflow()` by grouping and reordering
comparisons.

* Source/WebCore/rendering/RenderLayerModelObject.cpp:
(WebCore::RenderLayerModelObject::styleDidChange):
* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::RenderStyle::changeAffectsVisualOverflow const):
(WebCore::RenderStyle::scrollSnapDataEquivalent const):
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
(WebCore::RenderStyle::borderImageWidth const):
(WebCore::RenderStyle::contentData const):
(WebCore::RenderStyle::columnSpanEqual const):
(WebCore::RenderStyle::borderIsEquivalentForPainting const):
(WebCore::RenderStyle::contentDataEquivalent const):
(WebCore::RenderStyle::containerTypeAndNamesEqual const):
(WebCore::RenderStyle::scrollPaddingEqual const):
* Source/WebCore/rendering/style/StyleSurroundData.h:
* Source/WebCore/style/StyleChange.cpp:
(WebCore::Style::determineChange):
(WebCore::Style::operator<<):
* Source/WebCore/style/StyleChange.h:

Canonical link: https://commits.webkit.org/289700@main
